Add `DocumentSyncData` to process sync data machinery, and make first use of it for audio session

beidson opened the pull request:
#### 3c6899ca1a461716192c9dd3c2c0f7ce012c8f88
<pre>
Add `DocumentSyncData` to process sync data machinery, and make first use of it for audio session
<a href="https://rdar.apple.com/141294171">rdar://141294171</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284474">https://bugs.webkit.org/show_bug.cgi?id=284474</a>

Reviewed by Alex Christensen.

This patch adds a `DocumentSyncData` object - to be stored on Page - that is a stand-in for all
data that used to be stored on the topDocument().

It adds capabilities to generate-process-sync-data.py to enable:
- Data types to opt-in to being part of &quot;DocumentSyncData&quot;
- Data types to have a compile-time conditional
- Data types to specify their associated required header instead of throwing it into a global set

Most of the magic is in the generated code, as evidence by how little of the hand written code
had to change to support this.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
* Source/WebCore/Scripts/generate-process-sync-data.py:
(SyncedData.__init__):
(sorted_headers_from_datas):
(parse_process_sync_data):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(sorted_qualified_types):
(generate_process_sync_data_header):
(generate_document_synched_data_header):
(generate_document_synched_data_impl):
(generate_process_sync_data_serialiation_in):
(main):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
(WebCore::Document::setAudioSessionType): Deleted.
(WebCore::Document::audioSessionType const): Deleted.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::documentURLDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::m_documentSyncData):
(WebCore::Page::setMainFrameURLInternal):
(WebCore::Page::setMainFrameURL):
(WebCore::Page::setAudioSessionType):
(WebCore::Page::audioSessionType const):
(WebCore::Page::updateProcessSyncData):
(WebCore::m_activeNowPlayingSessionUpdateTimer): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/287717@main">https://commits.webkit.org/287717@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/decc51fcb2af60a61b9ce9467627a840d5d00296

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80604 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59611 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85128 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7917 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/62968 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73378 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43273 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50371 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27538 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30045 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28061 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86561 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7830 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5521 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8005 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69215 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/70502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17568 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14507 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13455 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7792 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7631 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11150 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9436 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
```diff
@@ -105,7 +110,12 @@ DOMAudioSession::Type DOMAudioSession::type() const
         return DOMAudioSession::Type::Auto;
 
-    return document ? document->topDocument().audioSessionType() : DOMAudioSession::Type::Auto;
+    if (document) {
```
If there's no document, this could just early return immediately instead of checking it twice and indenting code.

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -791,17 +793,39 @@ void Page::setMainFrame(Ref<Frame>&& frame)
 
+void Page::setMainFrameURL(const URL& url)
+{
+    setMainFrameURLInternal(url);
```
These could probably all have an early return if the thing your setting is equal to the thing that is already stored.

achristensen07 commented on Source/WebCore/Scripts/generate-process-sync-data.py:
```diff
@@ -54,7 +54,24 @@
+                if option == 'DocumentSyncData':
+                    self.automatic_location = "DocumentSyncData"
+                    continue
+                elif option.startswith('Conditional='):
```
It would be nice if this python had some tests that showed the generated cpp with a test .in file, kind of like "make -C Source/WebKit/Scripts/webkit/tests" re-generates cpp from .in files in that directory.

beidson replied:
I agree. I'll do that standalone after this patch.

beidson committed:
Add `DocumentSyncData` to process sync data machinery, and make first use of it for audio session
rdar://141294171
https://bugs.webkit.org/show_bug.cgi?id=284474

Reviewed by Alex Christensen.

This patch adds a `DocumentSyncData` object - to be stored on Page - that is a stand-in for all
data that used to be stored on the topDocument().

It adds capabilities to generate-process-sync-data.py to enable:
- Data types to opt-in to being part of "DocumentSyncData"
- Data types to have a compile-time conditional
- Data types to specify their associated required header instead of throwing it into a global set

Most of the magic is in the generated code, as evidence by how little of the hand written code
had to change to support this.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
* Source/WebCore/Scripts/generate-process-sync-data.py:
(SyncedData.__init__):
(sorted_headers_from_datas):
(parse_process_sync_data):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(sorted_qualified_types):
(generate_process_sync_data_header):
(generate_document_synched_data_header):
(generate_document_synched_data_impl):
(generate_process_sync_data_serialiation_in):
(main):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
(WebCore::Document::setAudioSessionType): Deleted.
(WebCore::Document::audioSessionType const): Deleted.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::documentURLDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::m_documentSyncData):
(WebCore::Page::setMainFrameURLInternal):
(WebCore::Page::setMainFrameURL):
(WebCore::Page::setAudioSessionType):
(WebCore::Page::audioSessionType const):
(WebCore::Page::updateProcessSyncData):
(WebCore::m_activeNowPlayingSessionUpdateTimer): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/287717@main
